lang/src/org/partiql/lang/eval/physical/operators/AggregateOperatorFactory.kt

```diff
@@ -0,0 +1,172 @@
+    ): AggregationFunctionFactory
+}
+
+internal class FullAggregateFactory(name: String) : AggregateOperatorFactory(name) {
+
+    private val aggregationFunctionFactories = mapOf(
+        AggregationFunctionFactoryKey(AggregationKind.MAX, "MAX") to MaxFactory("MAX"),
+        AggregationFunctionFactoryKey(AggregationKind.MIN, "MIN") to MinFactory("MIN"),
+        AggregationFunctionFactoryKey(AggregationKind.COUNT, "COUNT") to CountFactory("COUNT"),
+        AggregationFunctionFactoryKey(AggregationKind.SUM, "SUM") to SumFactory("SUM"),
+        AggregationFunctionFactoryKey(AggregationKind.GROUP_AS, "GROUP_AS") to GroupAsFactory("GROUP_AS"),
+        AggregationFunctionFactoryKey(AggregationKind.AVG, "AVG") to AverageFactory("AVG"),
+    )
+
+    override fun getFunctionFactory(operator: AggregationKind, name: String): AggregationFunctionFactory {
+        val key = AggregationFunctionFactoryKey(operator, name)
+        return aggregationFunctionFactories[key] ?: error("Factory for operator ${key.kind} named '${key.name}' does not exist.")
+    }
+
+    override fun create(
+        impl: PartiqlPhysical.Impl,
+        source: RelationExpression,
+        keys: List<CompiledGroupKey>,
+        functions: List<CompiledAggregateFunction>
+    ): RelationExpression =
+        RelationExpression { state ->
+            val sourceIter = source.evaluate(state)
+
+            val aggregationMap = AggregationMap()
+            val seen = List(functions.size) { TreeSet(DEFAULT_COMPARATOR) }
+            val registerSize = state.registers.size
+            val variables = functions.flatMap { function -> function.variables } + keys.map { key -> key.variable }
+            val variablesIndexes = variables.map { variable -> variable.index.value.toIntExact() }.toSet()
+
+            while (sourceIter.nextRow()) {
+
+                // Initialize the AggregationMap
+                val isMissingArray = Array(keys.size) { false }
+                val evaluatedGroupByKeys = keys.mapIndexed { index, groupByKey ->
+                    val ionValue = groupByKey.value.invoke(state).ionValue
+                    if (ionValue.isNullValue) {
+                        isMissingArray[index] = ionValue.hasTypeAnnotation(MISSING_ANNOTATION)
+                        ionValue.removeTypeAnnotation(MISSING_ANNOTATION)
+                    }
+                    ionValue
+                }
+                val groupMapValue = aggregationMap.getOrPut(evaluatedGroupByKeys) {
+                    GroupMapValue(
+                        state = EvaluatorState(
+                            state.session,
+                            state.valueFactory,
+                            Array(registerSize) { index ->
+                                when (variablesIndexes.contains(index)) {
+                                    true -> state.valueFactory.missingValue
+                                    false -> state.registers[index]
+                                }
```

Comment:
I am not sure I follow the logic here, mind explain a little more?

Reply:
Eh it might actually not be necessary anymore -- but I essentially only was initializing the output variable declarations with MISSING, but I can check if that's still necessary.